

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Exercise 4.2: Restriction enzyme cut sites &mdash; Programming Bootcamp  documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/copybutton.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/clipboard.min.js"></script>
        <script src="../../_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "document", "processHtmlClass": "math|output_area"}}</script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #F1F2D8" >
          

          
            <a href="../../index.html">
          

          
            
            <img src="../../_static/logo.png" class="logo" alt="Logo"/>
          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Lessons</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l00_configuring_your_computer.html">Lesson 0: Configuring your computer</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l01_welcome.html">Lesson 1: Welcome and Introduction to JupyterLab</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l02_basic_command_line_skills.html">Lesson 2: Basic command line skills</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l03_variables_operators_types.html">Lesson 3: Variables, operators, and types</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l04_more_operators_and_conditionals.html">Lesson 4: More operators and conditionals</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l05_lists_and_tuples.html">Lesson 5: Lists and tuples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l06_iteration.html">Lesson 6: Iteration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l07_intro_to_functions.html">Lesson 7: Introduction to functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l08_string_methods.html">Lesson 8: String methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l09_dictionaries.html">Lesson 9: Dictionaries</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l10_packages_and_modules.html">Lesson 10: Packages and modules</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l11_style.html">Lesson 11: Style</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l12_file_io.html">Lesson 12: File I/O</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l13_intro_to_oop.html">Lesson 13: Introduction to object-oriented programming</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l14_intro_to_pandas.html">Lesson 14: Introduction to Pandas</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l15_split_apply_combine.html">Lesson 15: Tidy data and split-apply-combine</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l16_plotting.html">Lesson 16: Making plots</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l17_intro_to_numpy_and_scipy.html">Lesson 17: Introduction to Numpy and Scipy</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l18_version_control_with_git.html">Lesson 18: Version control with Git</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l19_exceptions_and_error_handling.html">Lesson 19: Errors and Exception handling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l20_testing_and_tdd.html">Lesson 20: Testing and test-driven development</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l21_examples_of_tdd.html">Lesson 21: Examples of TDD</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l22_high_level_plotting.html">Lesson 20: High level plotting</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l23_plotting_time_series_generated_data.html">Lesson 22: Plotting time series and generated data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l24_intro_to_scripting.html">Lesson 24: Introduction to scripting</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l25_other_packages_and_languages.html">Lesson 25: Survey of other packages and languages</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../lessons/l26_bootcamp_recap.html">Lesson 26: Bootcamp recap</a></li>
</ul>
<p class="caption"><span class="caption-text">Schedule</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../schedule.html">Schedule overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../schedule.html#daily-schedule-all-times-pdt">Daily schedule (all times PDT)</a></li>
</ul>
<p class="caption"><span class="caption-text">Resources</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../resources.html">Scientific Python distribution</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../resources.html#online-instruction">Online instruction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../resources.html#books">Books</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Programming Bootcamp</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Exercise 4.2: Restriction enzyme cut sites</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../_sources/exercise_solutions/exercise_4/exercise_4.2_solution.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Exercise-4.2:-Restriction-enzyme-cut-sites">
<h1>Exercise 4.2: Restriction enzyme cut sites<a class="headerlink" href="#Exercise-4.2:-Restriction-enzyme-cut-sites" title="Permalink to this headline">¶</a></h1>
<hr><p><a class="reference external" href="https://en.wikipedia.org/wiki/Restriction_enzyme">Restriction enzymes</a> cut DNA at specific locations called <strong>restriction sites</strong>. The sequence at a restriction site is called a <strong>recognition sequence</strong>. Here are the recognition sequences of some commonly used restriction enzymes.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 72%" />
<col style="width: 28%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Restriction enzyme</p></th>
<th class="head"><p>Recognition sequence</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference external" href="https://en.wikipedia.org/wiki/HindIII">HindIII</a></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">AAGCTT</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://en.wikipedia.org/wiki/EcoRI">EcoRI</a></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">GAATTC</span></code></p></td>
</tr>
<tr class="row-even"><td><p>KpnI</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">GGTACC</span></code></p></td>
</tr>
</tbody>
</table>
<p><strong>a)</strong> <a class="reference external" href="https://www.neb.com/products/n3011-lambda-dna#Product%20Information">New England Biosystems</a> sells purified DNA of the genome of λ-phage, a bacteriophage that infect <em>E. coli</em>. You can download the FASTA file containing the sequence <a class="reference external" href="https://www.neb.com/-/media/nebus/page-images/tools-and-resources/interactive-tools/dna-sequences-and-maps/text-documents/lambdafsa.txt">here</a>. Use the function you wrote in <a class="reference external" href="exercise_4.1.ipynb">Exercise 3.1</a> to extract the sequence.</p>
<p><strong>b)</strong> Write a function with call signature</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">restriction_sites</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">recoq_seq</span><span class="p">)</span>
</pre></div>
</div>
<p>that takes as arguments a sequence and the recognition sequence of a restriction enzyme sites and returns the indices of the first base or each of the restriction sites in the sequence. Use this function to find the indices of the restriction sites of λ-DNA for HindIII, EcoRI, and KpnI. Compare your results to those reported on the <a class="reference external" href="https://www.neb.com/-/media/nebus/page-images/tools-and-resources/interactive-tools/dna-sequences-and-maps/lambda_sites.pdf">New England Biosystems datasheet</a>.</p>
<div class="section" id="Solution">
<h2>Solution<a class="headerlink" href="#Solution" title="Permalink to this headline">¶</a></h2>
<hr><div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">re</span>
</pre></div>
</div>
</div>
<p><strong>a)</strong> I have downloaded the FASTA file to <code class="docutils literal notranslate"><span class="pre">data/lambdaDNA.fasta</span></code>. Let’s load it in.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">def</span> <span class="nf">read_fasta</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Read a sequence in from a FASTA file containing a single sequence.</span>

<span class="sd">    We assume that the first line of the file is the descriptor and all</span>
<span class="sd">    subsequent lines are sequence.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="c1"># Read in descriptor</span>
        <span class="n">descriptor</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span>

        <span class="c1"># Read in sequence, stripping the whitespace from each line</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
        <span class="n">line</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span>
        <span class="k">while</span> <span class="n">line</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
            <span class="n">seq</span> <span class="o">+=</span> <span class="n">line</span>
            <span class="n">line</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">descriptor</span><span class="p">,</span> <span class="n">seq</span>


<span class="n">_</span><span class="p">,</span> <span class="n">seq</span> <span class="o">=</span> <span class="n">read_fasta</span><span class="p">(</span><span class="s1">&#39;data/lambdaDNA.fasta&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>We can check to make sure we got the whole sequence, which should be about 48.5 kbp.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
48502
</pre></div></div>
</div>
<p>Looks good!</p>
<p><strong>b)</strong> Our goal is to find all locations of the substring given by the recognition sequence in the genome. This is most easily accomplished using the <code class="docutils literal notranslate"><span class="pre">re</span></code> module that uses <a class="reference external" href="https://en.wikipedia.org/wiki/Regular_expression">regular expressions</a>, which we will not cover in the bootcamp. Specifically, we use <code class="docutils literal notranslate"><span class="pre">re.finditer()</span></code> to automatically find all occurrences of the recognition sequence in the sequence, and then use the <code class="docutils literal notranslate"><span class="pre">start()</span></code> method to get the first index.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">def</span> <span class="nf">restriction_sites_with_re</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">recog_seq</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Find the indices of all restriction sites in a sequence.&quot;&quot;&quot;</span>
    <span class="n">sites</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">site</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="n">recog_seq</span><span class="p">,</span> <span class="n">seq</span><span class="p">):</span>
        <span class="n">sites</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">site</span><span class="o">.</span><span class="n">start</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">sites</span>
</pre></div>
</div>
</div>
<p>Let’s check our restriction sites against the know sites.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;HindIII:&#39;</span><span class="p">,</span> <span class="n">restriction_sites_with_re</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s1">&#39;AAGCTT&#39;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;EcoRI:  &#39;</span><span class="p">,</span> <span class="n">restriction_sites_with_re</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s1">&#39;GAATTC&#39;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;KpnI:   &#39;</span><span class="p">,</span> <span class="n">restriction_sites_with_re</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s1">&#39;GGTACC&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
HindIII: [23129, 25156, 27478, 36894, 37458, 37583, 44140]
EcoRI:   [21225, 26103, 31746, 39167, 44971]
KpnI:    [17052, 18555]
</pre></div></div>
</div>
<p>Mostly, these match exactly the listed values, except each index reported by our function is one less than the listed values because we are indexing starting at zero. The one example that is different from the data sheet is that the data sheet does not contain the cut site at index 37583. This could be because NEB made a mistake either in the sequencing or in determining their cut sites, or because they validated the cut sites experimentally, and for some reason there is some strange structure
around that cut site that precludes cutting. It could also be that because there is only a short fragment of DNA between the cut site at 37458 and the one they missed at 37583 that is it missed in a gel electrophoresis assay.</p>
<p>I will now write a function do to the same calculation without using regular expressions. We will simply make a pass through the sequence and store the indices where we match the recognition sequence.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">def</span> <span class="nf">restriction_sites</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">recog_seq</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Find the indices of all restriction sites in a sequence.&quot;&quot;&quot;</span>
    <span class="c1"># Initialize list of restriction sites</span>
    <span class="n">sites</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Check every substring for a match</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">recog_seq</span><span class="p">)):</span>
        <span class="k">if</span> <span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="nb">len</span><span class="p">(</span><span class="n">recog_seq</span><span class="p">)]</span> <span class="o">==</span> <span class="n">recog_seq</span><span class="p">:</span>
            <span class="n">sites</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">sites</span>
</pre></div>
</div>
</div>
<p>And let’s use this function.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;HindIII:&#39;</span><span class="p">,</span> <span class="n">restriction_sites</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s1">&#39;AAGCTT&#39;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;EcoRI:  &#39;</span><span class="p">,</span> <span class="n">restriction_sites</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s1">&#39;GAATTC&#39;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;KpnI:   &#39;</span><span class="p">,</span> <span class="n">restriction_sites</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="s1">&#39;GGTACC&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
HindIII: [23129, 25156, 27478, 36894, 37458, 37583, 44140]
EcoRI:   [21225, 26103, 31746, 39167, 44971]
KpnI:    [17052, 18555]
</pre></div></div>
</div>
<p>For fun, we can compare the speeds of the two functions using the magic function <code class="docutils literal notranslate"><span class="pre">%timeit</span></code>. This function performs a calculation many times and computes a mean execution time. It can be used to check speed of your functions.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%</span><span class="k">timeit</span> restriction_sites(seq, &#39;AAGCTT&#39;)
<span class="o">%</span><span class="k">timeit</span> restriction_sites_with_re(seq, &#39;AAGCTT&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
9.03 ms ± 44.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
165 µs ± 2.75 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
</pre></div></div>
</div>
<p>The regular expression-based method is nearly 100 times faster! This is often the case; hand-coding something in pure Python can be slow compared to using packages that use pre-compiled code like <code class="docutils literal notranslate"><span class="pre">re</span></code>. However, as <a class="reference external" href="https://en.wikipedia.org/wiki/Donald_Knuth">Donald Knuth</a> said, “The real problem is that programmers have spent far too much time worrying about efficiency in the wrong places and at the wrong times; <strong>premature optimization is the root of all evil (or at least most of it) in
programming.</strong>” Get your code working, even if it is slow, and optimize only if you have to.</p>
</div>
<div class="section" id="Computing-environment">
<h2>Computing environment<a class="headerlink" href="#Computing-environment" title="Permalink to this headline">¶</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="o">%</span><span class="k">load_ext</span> watermark
<span class="o">%</span><span class="k">watermark</span> -v -p re,jupyterlab
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
CPython 3.7.7
IPython 7.16.1

re 2.2.1
jupyterlab 2.1.5
</pre></div></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
      <span class="lastupdated">
        Last updated on Aug 19, 2021.
      </span>

    </p>
  </div> 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
    <!-- Theme Analytics -->
    <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-65442910-1', 'auto');
    
    ga('send', 'pageview');
    </script>

    
   

</body>
</html>